crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2002-2012 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.crypto.scrypt;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.security.crypto.codec.Hex;
+import org.springframework.security.crypto.codec.Utf8;
+import org.springframework.security.crypto.keygen.KeyGenerators;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.bouncycastle.crypto.generators.SCrypt;
+
+
+/**
+ * Implementation of PasswordEncoder that uses the SCrypt hashing function. Clients
+ * can optionally supply a cpu cost parameter, a memory cost parameter and a parallelization parameter.
+ * 
+ * @author Shazin Sadakath
+ *
+ */
+public class SCryptPasswordEncoder implements PasswordEncoder {
+    
+    private final Log logger = LogFactory.getLog(getClass());
+    
+    private final int cpuCost;
+    
+    private final int memoryCost;
+    
+    private final int parallelization;
+    
+    private final byte[] saltBytes;
+    
+    public SCryptPasswordEncoder() {
+        this(new String(KeyGenerators.secureRandom().generateKey()), 16384, 8, 1);
+    }
+    
+    /**
+     * @param salt value for the algorithm
+     * @param cpu cost of the algorithm. must be power of 2 greater than 1
+     * @param memory cost of the algorithm
+     * @param parallelization of the algorithm
+     */
+    public SCryptPasswordEncoder(CharSequence salt, int cpuCost, int memoryCost, int parallelization) {

```

Comment:
Salt should be secure random generated for each individual password. That means we should not pass it into the constructor. Instead we should allow for a salt length to be injected and the salt should be added as a prefix of the password. Something like:

<salt><password>

we can then verify the password by extracting the salt using the number of bytes to extract out the salt that was generated.